src/codegen/compile.ml

```diff
@@ -808,56 +809,73 @@ end
 
 (* Function called compile_* return a list of instructions (and maybe other stuff) *)
 
-let compile_unboxed_const i = G.i (Const (nr (Wasm.Values.I32 i)))
-let compile_const_64 i = G.i (Const (nr (Wasm.Values.I64 i)))
-let compile_unboxed_zero = compile_unboxed_const 0l
-let compile_unboxed_one = compile_unboxed_const 1l
+
+let compile_comparison rel = 
+  G.i (Compare (Wasm_exts.Values.I64 rel)) ^^
+  G.i (Convert (Wasm_exts.Values.I64 I64Op.ExtendUI32))
+let compile_comparison_32 rel = 
+  G.i (Compare (Wasm_exts.Values.I32 rel)) ^^
+  G.i (Convert (Wasm_exts.Values.I64 I64Op.ExtendUI32))
+let compile_test op =
+  G.i (Test (Wasm_exts.Values.I64 op)) ^^
+  G.i (Convert (Wasm_exts.Values.I64 I64Op.ExtendUI32))
+let compile_comparison_f64 rel = 
+  G.i (Compare (Wasm_exts.Values.F64 rel)) ^^
+  G.i (Convert (Wasm_exts.Values.I64 I64Op.ExtendUI32))
```

Comment:
Curious: since G.if expects an i32, and the comparisons produce i32s, why widen to i64 only to narrow to i32 before  conditional branches (see prepare_branch below (also requiring the new peephole optimization in instrList.ml).

Reply:
The reason was simplicity because certain comparison are not immediately followed by a branch but some other operations like a logical or/and etc. Otherwise, I would follow the x86-architecture style of designing comparisons that are followed by branches (leaving the 32-bit comparison value) and use other comparisons whenever the result is processed further. I initially implemented the backend by keeping `i32` for boolean results (comparisons) but my impression was that the design became to complicated. Eventually, I thought, the solution with the peephole optimizer would make it more uniform.

Reply:
Ok, let's stick with this then.

Reply:
(i.e. continue to use 64-bit booleans).